File: solutions/089.Gray_Code/AC_simulation_2^n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_simulation_2^n.java
*  Create Date: 2015-02-20 10:30:33
*  Descripton:   
*/

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> res = new ArrayList<Integer>();
        for (int i = 0; i < (1<<n); ++i) {
            res.add(i ^ (i>>1));
        }
        return res;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 1};
        System.out.println(s.grayCode(2));
    }
}

File: solutions/089.Gray_Code/AC_simulation_2^n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_simulation_2^n.cpp
*  Create Date: 2015-02-20 10:00:25
*  Descripton:  simulation, G(n) = B(n) XOR B(n>>1)
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    vector<int> grayCode(int n) {
        const size_t sz = 1<<n;
        vector<int> res(sz);
        for (size_t i = 0; i < sz; ++i)
            res[i] = (i ^ (i>>1));
        return res;
    }
};

int main() {
    int n;
    Solution s;
    while (cin >> n) {
        vector<int> res = s.grayCode(n);
        for (auto &i : res)
            cout << i << ' ';
        cout << endl;
    }
    return 0;
}

